src/data/fin/tuple/sort.lean

```diff
@@ -91,3 +91,68 @@ begin
 end
 
 end tuple
+
+namespace tuple
+
+open list
+
+variables {n : ℕ} {α : Type*}
+
+/-- If we swap two strictly decreasing entries in a tuple, then the result is lexicographically
+smaller than the original tuple. -/
+lemma lex_desc [preorder α] {f : fin n → α} {i j : fin n} (h₁ : i < j) (h₂ : f j < f i) :
+  to_lex (f ∘ equiv.swap i j) < to_lex f :=
+begin
+  simp only [has_lt.lt, pi.lex, pi.to_lex_apply, function.comp_app],
+  refine ⟨i, λ k (hik : k < i), _, _⟩,
+  { rw [equiv.swap_apply_of_ne_of_ne (ne_of_lt hik) (ne_of_lt $ hik.trans h₁)], },
+  { simpa only [equiv.swap_apply_left], }
+end
+
+/-- If a tuple `f` and a permutation of `f` are both monotone, then they are equal. -/
+lemma unique_monotone [partial_order α] {f : fin n → α} {σ : equiv.perm (fin n)} (hf : monotone f)
+  (hfσ : monotone (f ∘ σ)) : f ∘ σ = f :=
+of_fn_injective $ eq_of_perm_of_sorted (σ.of_fn_comp_perm f) hfσ.of_fn_sorted hf.of_fn_sorted
+
+variables [linear_order α] {f : fin n → α}
+
+/-- A tuple `f` equals its sorted version if and only if it is monotone. -/
+lemma eq_comp_sort_iff_monotone : f ∘ sort f = f ↔ monotone f :=
```

Comment:
Is the following also true?
```suggestion
  lemma sort_eq_refl_iff_monotone : sort f = equiv.refl ↔ monotone f :=
```

Reply:
Yes if `f` is injective

Reply:
Oh I think it's always true; `sort` always produce the lexigraphically smallest permutation that sorts the tuple.

Reply:
But this is not part of the specification of `tuple.sort`. So for it to make sense as a lemma, the specification should include that `sort f` is the lexicographically smallest permutation making `f` monotone. (And the proof will need to use the implementation...)

Reply:
I think that should be part of the specification, if its already the case that the implementation satisfies it.

Reply:
I have added the lemma now. You are welcome to golf the proofs.

Reply:
I have now added the following lemma:
```lean
/-- A permutation `σ` is `sort f` if and only if `f ∘ σ` is monotone and whenever `i < j`
and `f (σ i) = f (σ j)`, then `σ i < σ j`. -/
lemma eq_sort_iff : σ = sort f ↔ monotone (f ∘ σ) ∧ ∀ i j, i < j → f (σ i) = f (σ j) → σ i < σ j :=
```
which gives the precise specification of `sort f `. This then gives a simpler proof of `sort_eq_refl_iff_monotone`.
Junyan will surely be able to condense the new proofs to a few lines :smile:

Reply:
Nice! I was planning to work on it but you were quicker. This is a bit different from the characterization I had in mind (lexicographically minimal), but likely easier to use, so I have no objections. (My approach would be first the order on `equiv.perm` as an instance (using [linear_order.lift'](https://leanprover-community.github.io/mathlib_docs/order/basic.html#linear_order.lift') of [pi.lex.linear_order](https://leanprover-community.github.io/mathlib_docs/data/pi/lex.html#pi.lex.linear_order)). The product order doesn't make sense (it's a discrete order) and the lexicographic is the only sensible one. Then I would show `equiv.refl _` is the bottom element.)

Reply:
Yes, I have thought a bit about how to formulate the characterization and came to the conclusion that stating it in terms of the lexicographic order on permutations would likely be somewhat painful and (as you say) less convenient to use.
I've golfed the new bits a little, but I guess more can be done.